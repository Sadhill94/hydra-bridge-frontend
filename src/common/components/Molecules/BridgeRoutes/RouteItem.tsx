import React, { ReactNode } from "react";
import styled from "styled-components";

import Icon from "../../Icon/Icon";
import { ContainerCard } from "../../Atoms/Containers/Container";
import {
  StyledBridgeArrow,
  StyledBridgeNetwork,
  StyledBridgeRoute,
  StyledBridgeRouteAmount,
} from "./styles";
import { FakeButton } from "../../Atoms/Buttons/Button";

import { IconKeys } from "../../../commonTypes";
import { stakenetTheme as theme } from "../../../../shell/theme/stakenetTheme";
import { devicesUp } from "../../../../media";

const CustomFakeButton = styled(FakeButton)<{
  hasBoxShadow?: boolean;
}>`
  &:focus,
  &:focus-within {
    background-color: ${theme.colors.blue.darker};
  }
`;
export const RouteItemContainerCard = styled(ContainerCard)<{
  isSelected: boolean;
}>`
  min-height: 10rem;
  padding: 1rem;
  background-color: ${theme.colors.blue.darkest};
  ${(props) =>
    props.isSelected ? `background-color: ${theme.colors.blue.darker}` : ""};
  border-radius: ${theme.borderRadius.lg};
  position: relative;
  z-index: 3;

  &:hover,
  &:focus {
    background-color: ${theme.colors.blue.darker};
  }
  @media only screen and ${devicesUp.md} {
    padding: 1.4rem 2.4rem;
  }
`;

export type RouteItemProps = {
  coinSymbol: IconKeys;
  bridgeSymbol: IconKeys;
  amountIn: string | number;
  amountOut: string | number;
  bridgeDisplayName: string;
  routeId: number;
  isSelected: boolean;
  onRouteSelect: (id: number) => void;
  children?: ReactNode;
};

const RouteItem = ({
  coinSymbol,
  bridgeSymbol,
  amountIn,
  amountOut,
  bridgeDisplayName,
  routeId,
  isSelected,
  onRouteSelect,
  children,
}: RouteItemProps) => {
  return (
    <CustomFakeButton
      ariaLabel={"select route"}
      hasBoxShadow={true}
      onClick={() => onRouteSelect(routeId)}
    >
      <RouteItemContainerCard isSelected={isSelected}>
        <StyledBridgeRoute>
          <StyledBridgeRouteAmount>
            <Icon
              name={coinSymbol}
              width={"2.8rem"}
              height={"3.2rem"}
              className={"amount__icon"}
            />
            <span className={"amount__number"}>{amountIn}</span>
          </StyledBridgeRouteAmount>
          <StyledBridgeArrow>
            <Icon
              name={"doubleArrowRight"}
              width={"2.1rem"}
              height={"1.8rem"}
              className={"arrow__icon"}
            />
          </StyledBridgeArrow>
          <StyledBridgeNetwork>
            <Icon
              name={bridgeSymbol}
              width={"2.6rem"}
              height={"2.3rem"}
              className={"network__icon"}
            />
            <p className={"network__name"}>{bridgeDisplayName}</p>
          </StyledBridgeNetwork>
          <StyledBridgeArrow>
            <Icon
              name={"doubleArrowRight"}
              width={"2.1rem"}
              height={"1.8rem"}
              className={"arrow__icon"}
            />
          </StyledBridgeArrow>
          <StyledBridgeRouteAmount rtl={true}>
            <Icon
              name={coinSymbol}
              width={"2.8rem"}
              height={"3.2rem"}
              className={"amount__icon"}
            />
            <span className={"amount__number"}>{amountOut}</span>
          </StyledBridgeRouteAmount>
        </StyledBridgeRoute>
        {children}
      </RouteItemContainerCard>
    </CustomFakeButton>
  );
};

export default RouteItem;
