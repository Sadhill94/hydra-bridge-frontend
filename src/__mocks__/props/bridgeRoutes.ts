import { BridgeRouteItemProps } from "../../common/components/Molecules/BridgeRoutes/RouteItem";

export const routeItemBaseProps: BridgeRouteItemProps = {
  amountIn: "0.0001",
  amountOut: "0.0001",
  coinSymbol: "ethCoin",
  bridgeSymbol: "polygonBridge",
  bridgeDisplayName: "Polygon",
  routeId: 2,
  isSelected: false,
  onRouteSelect: (id: number) => alert(`on route select id: ${id}`),
};
