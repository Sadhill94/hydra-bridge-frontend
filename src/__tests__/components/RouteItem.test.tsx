import React from "react";
import { cleanup, render, screen } from "@testing-library/react";

import RouteItem, {
  RouteItemProps,
} from "../../common/components/Molecules/BridgeRoutes/RouteItem";

import { mockRouteItemBaseProps } from "../../__mocks__/props/bridgeRoutes";
import { stakenetTheme as theme } from "../../shell/theme/stakenetTheme";

describe("The Bridge route item", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render correctly", async () => {
    const props: RouteItemProps = mockRouteItemBaseProps;

    const { asFragment } = render(<RouteItem {...props} />);
    expect(asFragment()).toMatchSnapshot();
    render(<RouteItem {...props} />);
  });

  it("should not be selected", async () => {
    const props: RouteItemProps = {
      ...mockRouteItemBaseProps,
    };
    render(<RouteItem {...props} />);
    const element = screen.getByRole("button");

    expect(element).toBeInTheDocument();
    expect(element.firstChild).toHaveStyle(
      `background-color: ${theme.colors.blue.darkest}`
    );
  });

  it("should be selected", async () => {
    const props: RouteItemProps = {
      ...mockRouteItemBaseProps,
      isSelected: true,
    };
    render(<RouteItem {...props} />);
    const element = screen.getByRole("button");

    expect(element).toBeInTheDocument();
    expect(element.firstChild).toHaveStyle(
      `background-color: ${theme.colors.blue.darker}`
    );
  });
});
