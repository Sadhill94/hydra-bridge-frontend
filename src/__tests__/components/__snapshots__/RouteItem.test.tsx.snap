// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Bridge route item should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    aria-label="select route"
    class="sc-iqseJM bzJIfm sc-crHmcD gYEbqi"
    role="button"
    tabindex="0"
  >
    <div
      class="sc-bdvvtL sc-jRQBWg fGXmcv fQFNvP sc-egiyK ixpTqd"
    >
      <div
        class="sc-gKclnd konHGI"
      >
        <div
          class="sc-iCfMLu gYGgwA"
        >
          <svg
            class="amount__icon"
            color="#FFFFFF"
            data-icon="ethCoin"
            height="3.2rem"
            width="2.8rem"
          >
            eth.svg
          </svg>
          <span
            class="amount__number"
          >
            0.0001
          </span>
        </div>
        <div
          class="sc-furwcr fhgROL"
        >
          <svg
            class="arrow__icon"
            color="#FFFFFF"
            data-icon="doubleArrowRight"
            height="1.8rem"
            width="2.1rem"
          >
            double-arrow-right.svg
          </svg>
        </div>
        <div
          class="sc-pVTFL qWzKy"
        >
          <svg
            class="chain__icon chain__icon--sm"
            color="#FFFFFF"
            data-icon="polygonBridge"
            height="2.6rem"
            width="2.6rem"
          >
            polygon.svg
          </svg>
          <p
            class="chain__name"
          >
            Polygon
          </p>
          <div
            class="chain__group"
          >
            <span>
              <svg
                class="network__icon"
                color="#FFFFFF"
                data-icon="polygonBridge"
                height="2.6rem"
                width="2.6rem"
              >
                polygon.svg
              </svg>
            </span>
            <span>
              Polygon
            </span>
          </div>
        </div>
        <div
          class="sc-furwcr fhgROL"
        >
          <svg
            class="arrow__icon"
            color="#FFFFFF"
            data-icon="doubleArrowRight"
            height="1.8rem"
            width="2.1rem"
          >
            double-arrow-right.svg
          </svg>
        </div>
        <div
          class="sc-iCfMLu fRazBy"
        >
          <svg
            class="amount__icon"
            color="#FFFFFF"
            data-icon="ethCoin"
            height="3.2rem"
            width="2.8rem"
          >
            eth.svg
          </svg>
          <span
            class="amount__number"
          >
            0.0001
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
